core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java

```diff
@@ -2927,4 +3018,96 @@ private static BsonDocument withNewVersion(
+        + "event"),
+    PENDING_WRITE_DELETE_MESSAGE("remote delete but a write is pending"),
+    PENDING_WRITE_EMPTY_VERSION_MESSAGE("remote or local have an empty version but a write is "
+        + "pending"),
+    PENDING_WRITE_INSTANCE_ID_MISMATCH_MESSAGE("pending write and remote events created by "
+        + "different devices"),
+    PROBABLY_GENERATED_BY_US_MESSAGE("remote change event was generated by us"),
+    REMOTE_UPDATE_WITHOUT_VERSION_MESSAGE("remote document changed but version was unmodified"),
+    SIMULTANEOUS_WRITES_MESSAGE("has multiple events at same logical time"),
+    STALE_LOCAL_WRITE_MESSAGE("remote event version has higher counter than local pending write"),
+    STALE_EVENT_MESSAGE("remote change event is stale"),
+    UNKNOWN_OPTYPE_MESSAGE("unknown operation type: %s"),
+    UNKNOWN_REMOTE_PROTOCOL_VERSION_MESSAGE("got a remote document with an unsupported "
+        + "synchronization protocol version %d"),
+    VERSION_DIFFERENT_DELETED_DOC_MESSAGE("version different on removed document"),
+    VERSION_DIFFERENT_REPLACED_DOC_MESSAGE("version different on replaced document or document was "
+        + "deleted"),
+    VERSION_DIFFERENT_UNPROCESSED_EVENT_MESSAGE("version different on unprocessed change event for "
+        + "document"),
+    VERSION_DIFFERENT_UPDATED_DOC_MESSAGE("version different on updated document or document was "
+        + "deleted");
+
+
+    static final String BASE_MESSAGE = "t='%d': %s ns=%s documentId=%s ";
+    static final String R2L_METHOD = "syncRemoteChangeEventToLocal";
+    static final String L2R_METHOD = "syncLocalToRemote";
+
+    final String message;
+
+    SyncMessage(final String message) {
+      this.message = BASE_MESSAGE.concat(message);
+    }
+
+    static String constructMessage(
+        final SyncAction action,
+        final SyncMessage message,
+        final Exception ex) {
+      final String messageWithDescription = action.applyDescription(message);
+
+      if (ex != null) {
+        return String.format(messageWithDescription, ex);
+      }
+
+      return messageWithDescription;
+    }
+
+    @Override
+    public String toString() {
+      return this.message;
+    }
+  }
+
+  private enum SyncAction {
+    APPLY_FROM_REMOTE(false, "; applying the remote event"),
+    APPLY_AND_VERSION_FROM_REMOTE(false, "; applying changes from the remote document"),
+    CONFLICT(true, "; raising conflict"),
+    DELETE_AND_DESYNC(false, "; deleting and desyncing the document"),
+    DELETE_FROM_REMOTE(false, "; applying the remote delete"),
+    DROP(false, "; dropping the event"),
+    DROP_AND_PAUSE(false, "; dropping the event and pausing the document"),
+    DROP_AND_DESYNC(false, "; dropping the event and desyncing the document"),
+    WAIT(false, "; waiting until next pass"),
+    REMOTE_FIND(false, "; re-checking against remote collection"),
+    SYNTHETIC_CONFLICT(true, "; raising conflict");
+
+    private final boolean conflict;
```

Comment:
[nit] This field is never accessed. We should either remove it or explain why we should keep it.

Reply:
removed